dashboard/utils.py

```diff
@@ -654,3 +654,30 @@ def get_address_for_submission_client(address: Optional[str]) -> str:
         address = ray_address_to_api_server_url(address)
     logger.debug(f"Using API server address {address}.")
     return address
+
+
+def compose_state_message(death_info_dict: dict) -> Optional[str]:
+    """Compose node state message based on death information.
+
+    Args:
+        death_info_dict: the node_death field in GcsNodeInfo, in dict type.
+    """
+    death_reason = death_info_dict.get("reason", None)
+    if death_reason == "EXPECTED_TERMINATION":
+        state_message = "Expected termination"
+    elif death_reason == "UNEXPECTED_TERMINATION":
+        state_message = "Unexpected termination"
+    elif death_reason == "AUTOSCALER_DRAIN_PREEMPTED":
+        state_message = "Terminated due to preemption from cloud provider"
+    elif death_reason == "AUTOSCALER_DRAIN_IDLE":
+        state_message = "Terminated due to idle (no Ray activity)"
+    else:
+        state_message = None
+
+    death_reason_message = death_info_dict.get("reasonMessage", None)
+    if death_reason_message:
+        if state_message:
+            state_message += f": ({death_reason_message})"
+        else:
+            state_message = death_reason_message
+    return state_message
```

Comment:
Let's review the state message for each case and make sure it's good.

Reply:
- Expected termination: received SIGTERM
- Unexpected termination: health check failed due to missing too many heartbeats.
- Terminated due to idle (no Ray activity): idle for 5 seconds
- Terminated due to preemption: spot preemption
- Terminated due to preemption: on-demand to spot replacement

Reply:
Addressed at https://github.com/ray-project/ray/pull/45357/commits/70293912329a1df001287e1dd415751dd03c24cf